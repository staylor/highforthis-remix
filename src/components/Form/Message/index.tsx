import cn from 'classnames';
import type { SyntheticEvent } from 'react';
import { useSearchParams } from '@remix-run/react';

interface MessageProps {
  dismissable?: boolean;
  text?: string;
  param?: string;
}

export default function Message({
  dismissable = true,
  text = '',
  param = 'message',
}: MessageProps) {
  const [searchParams, setSearchParams] = useSearchParams();
  const currentParam = searchParams.get(param) || '';
  if (!text || (dismissable && !currentParam)) {
    return null;
  }
  let message = text;
  if (dismissable && text.includes('%s')) {
    message = text.replace('%s', currentParam);
  }
  const onClick = (e: SyntheticEvent) => {
    e.preventDefault();

    searchParams.delete(param);
    setSearchParams(searchParams);
  };

  return (
    <div
      className={cn(
        'border-pink border-l-4 bg-neutral-50 shadow-sm',
        'relative mb-4 mt-1 block py-px pr-9 pl-3'
      )}
    >
      <p className="my-2 p-0.5 text-sm text-dark">{message}</p>
      {dismissable && (
        <button
          type="button"
          className={cn(
            'm-0 cursor-pointer border-none bg-none p-2',
            'absolute right-1 top-0',
            'dashicons-before before:content-dismiss before:text-base',
            'before:block before:h-5 before:w-5 before:bg-none'
          )}
          onClick={onClick}
        />
      )}
    </div>
  );
}
