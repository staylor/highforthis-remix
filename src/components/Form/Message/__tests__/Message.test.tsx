import { describe, test, expect } from 'vitest';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';

import Message from '@/components/Form/Message';

describe('Message', () => {
  test('empty', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/foo?message=updated']}>
        <Message />
      </MemoryRouter>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('text', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/foo?message=updated']}>
        <Message text="This is an admin message." />
      </MemoryRouter>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
