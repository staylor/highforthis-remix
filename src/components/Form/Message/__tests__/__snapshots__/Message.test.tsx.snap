// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Message > empty 1`] = `null`;

exports[`Message > text 1`] = `
<div
  class="border-pink border-l-4 bg-neutral-50 shadow-sm relative mb-4 mt-1 block py-px pr-9 pl-3"
>
  <p
    class="text-dark my-2 p-0.5 text-sm"
  >
    This is an admin message.
  </p>
  <button
    class="m-0 cursor-pointer border-none bg-none p-2 absolute right-1 top-0 dashicons-before before:content-dismiss before:text-base before:block before:h-5 before:w-5 before:bg-none"
    type="button"
  />
</div>
`;
